### PR TITLE
feat(cla): add caicloud cla code

### PR DESCRIPTION
**Pull request by bbbmj**
add caicloud cla code to verify email
only validate caicloud members.

### CODE REVIEW

File: mungegithub/mungers/submit-queue.go [hunk 1]
```diff
@@ -60,8 +60,8 @@ const (
 	holdLabel                        = "do-not-merge/hold"
 	deprecatedReleaseNoteLabelNeeded = "release-note-label-needed"
 	releaseNoteLabelNeeded           = "do-not-merge/release-note-label-needed"
-	cncfClaYesLabel                  = "cncf-cla: yes"
-	cncfClaNoLabel                   = "cncf-cla: no"
+	cncfClaYesLabel                  = "caicloud-cla: yes"
```
Comment (deyuanix) [inline mungegithub/mungers/submit-queue.go]: s/cncfClaYesLabel/caicloudClaYesLabel?

File: prow/plugins/caicloud-cla/cla.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		log.WithError(err).Warningf("Unable to fetch PR-%d from %s/%s.", num, org, repo)
+		return nil
+	}
+
+	number := pr.Number
+	isMember, err := gc.IsMember(org, pr.User.Login)
+	if err != nil {
+		log.WithError(err).Errorf("Error from IsMember(%q of org %q).", pr.User.Login, org)
+	}
+	if !isMember {
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla.go]: the logic here is "if someone is not a member, we add `cla: yes` label"?  shouldn't we add `cla:no`?

File: prow/plugins/caicloud-cla/cla.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		return nil
+	}
+
+	commit, err := gc.GetCommit(org, repo, pr.Head.SHA)
+	if err != nil {
+		log.WithError(err).Warningf("Unable to get commit-%s from %s/%s.", pr.Head.SHA, org, repo)
+		return nil
+	}
+
+	wantCLA := false
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla.go]: This variable is confusing... `wantCLA` means we want to have the author to sign CLA, but the logic below conflicts with my impression... maybe this should be named `claReady` or something similar?

File: prow/plugins/caicloud-cla/cla.go (hunk 2 above)
Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla.go]: If someone is not a member, he won't have email with `caicloud.io` domain.
If we add `cla: no`, then, his pr can't be merged.

File: prow/plugins/caicloud-cla/cla.go (hunk 3 above)
Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla.go]: good idea of `claReady `!

**Comment by bbbmj** [comment]
PTAL @ddysher

File: prow/plugins/caicloud-cla/cla.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+	if hasCaicloudYes {
+		if err := gc.RemoveLabel(org, repo, number, claYesLabel); err != nil {
+			log.WithError(err).Warningf("Could not remove %s label.", claYesLabel)
+		}
+		if err := gc.AddLabel(org, repo, number, claNoLabel); err != nil {
+			log.WithError(err).Warningf("Could not add %s label.", claNoLabel)
+		}
+	}
+	if err := gc.CreateComment(org, repo, number, fmt.Sprintf(caicloudclaNotFoundMessage, plugins.AboutThisBot)); err != nil {
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla.go]: initially, a PR has no `hasCaicloudYes` and `hasCaicloudNo` (both will be false), based on your logic here,  even if i'm a caicloud member, i still get this cla not found message?

File: prow/plugins/caicloud-cla/cla.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName                 = "caicloud-cla"
+	claYesLabel                = "caicloud-cla: yes"
+	claNoLabel                 = "caicloud-cla: no"
+	caicloudclaNotFoundMessage = `Thanks for your pull request. Before we can look at your pull request, you'll need to finish a Contributor License Agreement (CLA).
+
+:memo: **Please follow instructions at <https://github.com/caicloud/engineering/blob/master/docs/CLA.md> to complete the CLA.**
```
Comment (supereagle) [inline prow/plugins/caicloud-cla/cla.go:20]: engineering repo is private, if the author is not caicloud member, he/she will can not open the CAL.md.

File: prow/plugins/caicloud-cla/cla.go (hunk 2 above)
Comment (supereagle) [reply prow/plugins/caicloud-cla/cla.go]: This should be considered in the design of CLA system. We need to store the non members have signed CLA.

File: prow/plugins/caicloud-cla/cla.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		}
+		if err := gc.AddLabel(org, repo, number, claNoLabel); err != nil {
+			log.WithError(err).Warningf("Could not add %s label.", claNoLabel)
+		}
+	}
+
+	if err := gc.CreateComment(org, repo, number, fmt.Sprintf(caicloudclaNotFoundMessage, plugins.AboutThisBot)); err != nil {
+		log.WithError(err).Warning("Could not create CLA not found comment.")
+	}
+	if err := gc.AddLabel(org, repo, number, claNoLabel); err != nil {
```
Comment (supereagle) [inline prow/plugins/caicloud-cla/cla.go:159]: `gc.AddLabel()` will be duplicately called if `hasCaicloudYes` is true. So this part should be move into `if hasCaicloudYes {...}`.

**Comment by deyuanix** [comment]
@bbbmj please finish this ASAP and also add unit tests.  The PR logic is a little twisted.

**Comment by bbbmj** [comment]
done @ddysher @supereagle

**Comment by bbbmj** [comment]
After pr merged，I will update the pod

File: prow/plugins/caicloud-cla/cla.go (hunk 5 above)
Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla.go:20]: if the author is not caicloud member, he won't get the comment.

Comment (supereagle) [reply prow/plugins/caicloud-cla/cla.go:20]: Then what he will get?

Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla.go:20]: Now that we don't have CLA system, if one is not caicloud member, he will get `caicloud-cla: yes` directly.

File: prow/plugins/caicloud-cla/cla.go (hunk 6 above)
Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla.go:159]: the logic here is changed.

File: prow/plugins/caicloud-cla/cla_test.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+			addedLabels:   []string{fmt.Sprintf("%s", claNoLabel)},
+			removedLabels: nil,
+		},
+		{
+			name:     "no label, is member, cla not ready (right author email)",
+			labels:   sets.NewString(),
+			isMember: true,
+			commit: &github.CommitData{
+				Author: github.Author{
+					Email: "bmj@caiclou.io",
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla_test.go]: s/caiclou/caicloud

File: prow/plugins/caicloud-cla/cla_test.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+		{
+			name:     "no label, is member, cla not ready (right committer email)",
+			labels:   sets.NewString(),
+			isMember: true,
+			commit: &github.CommitData{
+				Author: github.Author{
+					Email: "bmj@gmail.com",
+				},
+				Committer: github.Committer{
+					Email: "bmj@caiclou.io",
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla_test.go]: s/caiclou/caicloud

File: prow/plugins/caicloud-cla/cla_test.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+		err           error
+		createCommentErr, addLabelErr, removeLabelErr, getIssueLabelsErr, getCommitErr,
+		getPullRequestErr, isMemberErr error
+	}{
+		{
+			name:     "no label, not member",
+			labels:   sets.NewString(),
+			isMember: false,
+			commit: &github.CommitData{
+				Author: github.Author{
```
Comment (deyuanix) [inline prow/plugins/caicloud-cla/cla_test.go]: what's the difference between `author` and `committer`?

File: prow/plugins/caicloud-cla/cla_test.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+func (c *ghc) IsMember(_, _ string) (bool, error) {
+	c.T.Log("IsMember")
+	return c.isMember, c.isMemberErr
+}
+
+func TestHandlePR(t *testing.T) {
+	cases := []struct {
+		name          string
```
Comment (supereagle) [inline prow/plugins/caicloud-cla/cla_test.go]: Define the `cases` as a map, then move the name out of the struct, and make it as the key of the map to describe the test case.

```go
	cases := map[string]struct {
		labels        sets.String
		isMember      bool
		action        github.PullRequestEventAction
		commit        *github.CommitData
		pr            *github.PullRequest
		addedLabels   []string
		removedLabels []string
		err           error
		createCommentErr, addLabelErr, removeLabelErr, getIssueLabelsErr, getCommitErr,
		getPullRequestErr, isMemberErr error
	}{
		"no label, not member": {
			labels:   sets.NewString(),
			isMember: false,
			commit: &github.CommitData{
				Author: github.Author{
					Email: "bmj@gmail.com",
				},
				Committer: github.Committer{
					Email: "bmj@gmail.com",
				},
			},
			pr: &github.PullRequest{
				Number: 101,
			},
			action:        github.PullRequestActionOpened,
			addedLabels:   []string{fmt.Sprintf("%s", claYesLabel)},
			removedLabels: nil,
		},
	}
````

File: prow/plugins/caicloud-cla/cla_test.go (hunk 9 above)
Comment (bbbmj) [reply prow/plugins/caicloud-cla/cla_test.go]: we can see the definition here `https://developer.github.com/v3/git/commits/#optional-parameters`

**Comment by bbbmj** [comment]
Thank you all for careful review，PTAL @ddysher @supereagle

**Comment by deyuanix** [comment]
LGTM, manually merge.

**Pull request merged**